src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -1996,55 +2003,107 @@ static private void FlattenDeconstructVariables(ImmutableArray<DeconstructionVar
         /// </summary>
-        private static MethodSymbol FindDeconstruct(int numCheckedVariables, BoundExpression boundRHS, SyntaxNode node, DiagnosticBag diagnostics)
+        private BoundExpression MakeDeconstructInvocationExpression(
+                                    int numCheckedVariables, BoundExpression receiver, AssignmentExpressionSyntax assignmentSyntax,
+                                    DiagnosticBag diagnostics, out ImmutableArray<BoundDeconstructValuePlaceholder> outPlaceholders)
         {
-            // find symbol for Deconstruct member
-            ImmutableArray<Symbol> candidates = boundRHS.Type.GetMembers("Deconstruct");
-            switch (candidates.Length)
+            var receiverSyntax = receiver.Syntax;
+
+            if (receiver.Type.IsDynamic())
             {
-                case 0:
-                    Error(diagnostics, ErrorCode.ERR_MissingDeconstruct, boundRHS.Syntax, boundRHS.Type);
-                    return null;
-                case 1:
-                    break;
-                default:
-                    Error(diagnostics, ErrorCode.ERR_AmbiguousDeconstruct, boundRHS.Syntax, boundRHS.Type);
-                    return null;
+                Error(diagnostics, ErrorCode.ERR_CannotDeconstructDynamic, receiverSyntax);
+                outPlaceholders = default(ImmutableArray<BoundDeconstructValuePlaceholder>);
+
+                return BadExpression(receiverSyntax, receiver);
             }
 
-            Symbol deconstructMember = candidates[0];
+            var analyzedArguments = AnalyzedArguments.GetInstance();
+            var outVars = ArrayBuilder<OutDeconstructVarPendingInference>.GetInstance(numCheckedVariables);
+            DiagnosticBag bag = null;
 
-            // check that the deconstruct fits
-            if (deconstructMember.Kind != SymbolKind.Method)
+            try
             {
-                Error(diagnostics, ErrorCode.ERR_MissingDeconstruct, boundRHS.Syntax, boundRHS.Type);
-                return null;
-            }
+                for (int i = 0; i < numCheckedVariables; i++)
+                {
+                    var variable = new OutDeconstructVarPendingInference(assignmentSyntax);
+                    analyzedArguments.Arguments.Add(variable);
+                    analyzedArguments.RefKinds.Add(RefKind.Out);
+                    outVars.Add(variable);
+                }
 
-            MethodSymbol deconstructMethod = (MethodSymbol)deconstructMember;
-            if (deconstructMethod.MethodKind != MethodKind.Ordinary)
-            {
-                Error(diagnostics, ErrorCode.ERR_MissingDeconstruct, boundRHS.Syntax, boundRHS.Type);
-                return null;
-            }
+                string methodName = "Deconstruct";
+                var memberAccess = BindInstanceMemberAccess(
+                                        receiverSyntax, receiverSyntax, receiver, methodName, rightArity: 0,
```

Comment:
Actually, the tests for deconstructing `System.Tuple` did involve generic extensions. I added an explicit test though.